Bootloop detection & recovery

DedeHai opened the pull request:
- Add functions to detect and handle bootloops
- Add functions to handle file backup and restore
- Add some wrappers and function to dump .json files to serial

How boot loop detection works:
All ESPs have an internal timer that is persistent: this is used to determine intervals between boots. If the reported reset reason hints at a crash, a persistent counter is incremented. If this happens several times in a row, it’s boot looping.
To recover from a boot loop three steps are attempted
1. Load config from backup (backup is created before every config save)
2. Reset config by renaming the file: defaults are applied at next boot (different name than normal backup)
3. Swap to the other boot partition if possible (rollback)
4. If all of the above failed: keeps boot-looping but dumps all .json files to serial

Brownout detection
Bootloops due to brownout cannot be detected with the above method unless resorting to flash-stored variables. While possible, it is more complicated. The simple solution: if a brownout is detected, the config is restored from last backup: this will fix any immediate crashes after saving a bad choice. It will however also revert the last change if a brownout happens due to some other reason like setting the LEDs too bright but can’t account for hardware deficiency of user setups.

Backup are intentionally ending in .json so they can be displayed in file editor. Not sure this is the best idea: users could also download the file to view it, inexperienced users should not mess with the backups: should rather add a UI element to restore backups.

Ideas / additions based on this PR:
- Add a button to the welcome page if a config backup is found to „restore config from backup“
- Automatic backup of presets: add a variable „presetsNeedBackup“, save presets to a backup file next time the LEDs are off (large preset files cause delay if done at *runtime*). downside: reduces number of presets that fit in FS, so this may need a setting option to disable
- Add a serial command to invoke the .json dump


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Device-side backup, restore, verification, and reset of configuration files.
  * File utilities for copying, comparing, validating JSON, restoring backups, and dumping JSON files to serial.
  * Automatic bootloop detection and staged recovery (config restore/reset and boot partition rollback where supported).

* **Improvements**
  * Configurations are backed up before overwrite and during updates.
  * Startup now validates and attempts recovery of configuration before use.
  * Improved resilience with automated multi-step recovery on repeated boot failures.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Arcitec commented:
Wow, this idea is seriously **awesome** and will be a huge benefit for people who don't have a serial/USB port for flashing. Thank you for doing this. This will definitely save a lot of people who only have OTA / WiFi on their boards. I love everything about this!

I also really love that you're solving it via the ESP32's RTC which persists between reboots. That's smart! The 5-second reboot intervals detected 5 times in a row looks good, and you're clearing the crash-counter on a successful boot. The attempted auto-recovery methods also look great with the config restore, then a config reset, and finally boot partition swap.

Great job, man. Very clever stuff! :heart:

DedeHai commented on wled00/file.cpp:
```diff
@@ -439,3 +439,96 @@
+
+static const char s_backup_json[] PROGMEM = ".bkp.json";
+
+bool backupFile(const char* filename) {
+  DEBUG_PRINTF("backup %s", filename);
+  if (!WLED_FS.exists(filename)) return false;
+  char backupname[32];
+  strcpy(backupname, filename);
+  strcat(backupname, s_backup_json);
+
+  if (copyFile(filename, backupname)) {
+    DEBUG_PRINTLN(F("backup ok"));
+    return true;
+  }
+  DEBUG_PRINTLN(F("backup failed"));
+  return false;
+}
+
+bool restoreFile(const char* filename) {
+  DEBUG_PRINTF("restore %s", filename);
+  char backupname[32];
+  strcpy(backupname, filename);
+  strcat(backupname, s_backup_json);
+
+  if (!WLED_FS.exists(backupname)) {
+    DEBUG_PRINTLN(F("no backup found"));
+    return false;
+  }
+
+  if (copyFile(backupname, filename)) {
+    DEBUG_PRINTLN(F("restore ok"));
+    return true;
+  }
+  DEBUG_PRINTLN(F("restore failed"));
+  return false;
+}
```
this is true, however, the backed-up file names are short and are under control of the devs. choosing long file names for essential files that need backup is very highly unlikely.

DedeHai commented:
Fixed some minor issues in the last commit.
I also removed reverting the config on brown-out detection: I was not able to test that as my ESP32 just never would trigger a brownout and I think blindly reverting the config can cause more trouble than its worth.

willmmiles reviewed (approved):
Looks great to me!  I'd still suggest maybe ".backup.json", but I don't think it matters that much.  I don't think there's any down side to merging this.

DedeHai commented:
thanks for the review. I think I have some uncommited work on this PR, one change is to use bak.filename.json as suggested.

netmindz commented:
I think this feature might be worth including in the next maintenance release

DedeHai commented:
> I think this feature might be worth including in the next maintenance release

not 100% sure it is 1:1 compatible with V3 builds, some API calls may have been renamed, a port is certainly possible if that is the case.
